src/ioctls.rs

```diff
@@ -0,0 +1,157 @@
+    length: u64,
+    secinfo: u64,
+    flags: u64,
+    count: u64,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> AddPages<'a> {
+    /// Create a new instance.
+    pub fn new(bytes: &'a [u8], offset: usize, secinfo: &'a SecInfo, measure: bool) -> Self {
+        const MEASURE: u64 = 1 << 0;
+
+        let flags = match measure {
+            true => MEASURE,
+            false => 0,
+        };
+
+        Self {
+            src: bytes.as_ptr() as _,
+            offset: offset as _,
+            length: bytes.len() as _,
+            secinfo: secinfo as *const _ as _,
+            flags,
+            count: 0,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Read the count attribute.
+    pub fn count(&self) -> u64 {
+        self.count
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+/// Parameters for ENCLAVE_INIT.
+pub struct Init<'a>(u64, PhantomData<&'a ()>);
+
+impl<'a> Init<'a> {
+    /// Create a new instance.
+    pub fn new(sig: &'a Signature) -> Self {
+        Init(sig as *const _ as _, PhantomData)
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+/// Parameters for ENCLAVE_RESTRICT_PERMISSIONS.
+pub struct RestrictPermissions<'a> {
+    /// In: starting page offset
+    offset: u64,
+    /// In: length of the address range (multiple of the page size)
+    length: u64,
+    /// In: SECINFO containing the relaxed permissions
+    secinfo: u64,
+    /// Out: ENCLU[EMODPR] return value
+    result: u64,
+    /// Out: length of the address range successfully changed
+    count: u64,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> RestrictPermissions<'a> {
+    /// Create a new RestrictPermissions instance.
+    pub fn new(offset: usize, length: usize, secinfo: &'a SecInfo) -> Self {
+        Self {
+            offset: offset as _,
+            length: length as _,
+            secinfo: secinfo as *const _ as _,
+            result: 0,
+            count: 0,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Read the result attribute.
+    pub fn result(&self) -> u64 {
+        self.count
+    }
+
+    /// Read the count attribute.
+    pub fn count(&self) -> u64 {
+        self.count
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::page::Flags;
+    use std::fs::OpenOptions;
+
+    #[test]
+    fn restrict_permissions() {
+        let mut device_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open("/dev/sgx_enclave")
+            .unwrap();
+
+        let secinfo = SecInfo::reg(Flags::empty());
+        let mut parameters = RestrictPermissions::new(0, 0, &secinfo);
+
+        let ret = match ENCLAVE_RESTRICT_PERMISSIONS.ioctl(&mut device_file, &mut parameters) {
+            Ok(_) => 0,
+            Err(err) => err.raw_os_error().unwrap(),
+        };
+
+        assert!(match ret {
+            // SGX1
+            libc::ENOTTY => true,
+            // SGX2
+            libc::EINVAL => true,
+            _ => false,
+        })
```

Comment:
This is the only place where `libc` is used. Also, this is only usable on x86_64-linux. So just hard code the two values here and drop the `libc` dependency. It isn't worth the weight of that crate for two consts.

So put these in the `tests` module (out of the function scope):
```rust
const EINVAL: i32 = 22;
const ENOTTY: i32 = 25;
```

Then do this:
```suggestion
assert!(ret == ENOTTY || ret == EINVAL); 
```

Reply:
Yeah, 2nd one is funny, started to trip too much into Rust I guess.